File: redfox-app/src/main/kotlin/io/github/malczuuu/redfox/app/core/ThingService.kt
package io.github.malczuuu.redfox.app.core

import io.github.malczuuu.redfox.app.common.Identity
import io.github.malczuuu.redfox.app.common.PageResult
import io.github.malczuuu.redfox.app.domain.ProjectRepository
import io.github.malczuuu.redfox.app.domain.ThingEntity
import io.github.malczuuu.redfox.app.domain.ThingRepository
import java.time.Clock
import java.time.Instant
import java.util.UUID
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ThingService(
    private val projectRepository: ProjectRepository,
    private val thingRepository: ThingRepository,
    private val clock: Clock,
) {

  fun getThings(projectId: UUID, pageable: Pageable): PageResult<ThingDto> {
    requireExistingProject(projectId)
    val things = thingRepository.findAll(projectId, pageable)
    return PageResult(
        content = things.content.map { it.toDto() },
        page = things.number,
        size = things.size,
        totalElements = things.totalElements,
    )
  }

  fun getThing(projectId: UUID, id: UUID): ThingDto {
    requireExistingProject(projectId)
    val thing = thingRepository.findOne(projectId, id).orElseThrow { ThingNotFoundException() }
    return thing.toDto()
  }

  @Transactional
  fun createThing(projectId: UUID, request: CreateThingDto): Identity {
    val project = projectRepository.findOne(projectId).orElseThrow { ProjectNotFoundException() }
    if (thingRepository.existsByCode(projectId, request.code)) {
      throw ThingAlreadyExistsException()
    }
    var thing =
        ThingEntity(
            project = project,
            code = request.code,
            name = request.name,
            description = request.description,
        )
    thing = thingRepository.save(thing)
    return Identity(thing.id.toString())
  }

  @Transactional
  fun updateThing(projectId: UUID, id: UUID, request: UpdateThingDto) {
    requireExistingProject(projectId)
    val thing = thingRepository.lockOne(projectId, id).orElseThrow { ThingNotFoundException() }
    thing.name = request.name
    thing.description = request.description
    thingRepository.save(thing)
  }

  @Transactional
  fun deleteThing(projectId: UUID, id: UUID) {
    requireExistingProject(projectId)
    thingRepository.lockOne(projectId, id).ifPresent { thing ->
      thing.deletedAt = Instant.now(clock)
      thingRepository.save(thing)
    }
  }

  private fun requireExistingProject(projectId: UUID) {
    projectRepository.findOne(projectId).orElseThrow { ProjectNotFoundException() }
  }

  private fun ThingEntity.toDto() =
      ThingDto(
          id = id,
          code = code,
          name = name,
          description = description,
          createdAt = createdAt,
          updatedAt = updatedAt,
          version = version,
      )
}

File: redfox-webapp/src/app/core/thing/thing.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Identity, PageResult } from '../common/common.model';
import { CreateThing, Thing, UpdateThing } from './thing.model';

@Injectable({ providedIn: 'root' })
export class ThingService {
  private readonly baseUrl = '/api/v1/projects';

  constructor(private readonly http: HttpClient) {}

  getThings(projectId: string, page = 0, size = 20): Observable<PageResult<Thing>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<PageResult<Thing>>(`${this.baseUrl}/${projectId}/things`, { params });
  }

  getThing(projectId: string, id: string): Observable<Thing> {
    return this.http.get<Thing>(`${this.baseUrl}/${projectId}/things/${id}`);
  }

  createThing(projectId: string, request: CreateThing): Observable<Identity> {
    return this.http.post<Identity>(`${this.baseUrl}/${projectId}/things`, request);
  }

  updateThing(projectId: string, id: string, request: UpdateThing): Observable<void> {
    return this.http.put<void>(`${this.baseUrl}/${projectId}/things/${id}`, request);
  }

  deleteThing(projectId: string, id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${projectId}/things/${id}`);
  }
}
